### PR TITLE
Feature: several things for Spectrum preparation

### PR DESCRIPTION
**Pull request by k9ert**
In order to integrate Spectrum as a "node" into the system, we have several changes:
- We now need to distinguish between different kinds of nodes than just external or not. For that:
  - We create a new key in the node.js-files: `python_class` which contains the class which can be instantiated via its `fromJson` method
  - That has a Specter migration which takes care of the migration of this data persistence change. The rollback of this change is described in the Specter migration as well
  - This caused also some changes in the inheritance hierarchy, so now a `PersistentObject` is aware of its `python_class` and can search for the class and calls in its `from_json` method the `from_json` method of that specific class . An `AbstractNode` has a better idea which methods/properties are necessary and it has some non performant default-implementation which can be overridden in subclasses. The `SpectrumNode` is obviously not implemented here but in the related https://github.com/cryptoadvance/spectrum/pull/9:
  [](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNp9UsFOwzAM_ZUp4rCh7QcqLkhwBaTdUKUoTd3WaxsXx2WgsX8nXTeasWk-RHn2i_1ekp2ylINKlG2M909oSjZt6mYhxvUN2KMXcPKabcDK7OFntZo9Zl7YWHkJh0denBk5t2rrLrTivp04t2ZeMuL1oPzizG4sDlF8WDehjacIodd-0AKsLTFoigYOAV-iMZ9w1vTQMbqIUjC1eug6X9yN2X0sLDYfiUJX0IQcyJa41ufZvhNsYcK2Mhhpz1AsoRuFfw7-yWk228u2nQmvCsGkP3POvXPoyngA2FpnDdm6AiwrmS_ur1g6WrlSid_1P0MtVQvcGszDdzvcRKqkgmBQJWGbG65Tlbp94PVdbgSecxRilRSm8bBUphdafzurkjABTqTjl_1jdca9E53w_hc3VOeK)
  
- We now can extend certain pages via "ViewModels" (VMs). For Spectrum, we have the `WelcomeVm` we enables the changes in the about-page (which is by default shown if no node is configured). This causes the changes in the `server_endpoints/welcome` which is now a dedicated package where the `WelcomeVm` lives in. Also it's now adding a new callback in `services/callbacks`: `adjust_view_model`
- The `WelcomeVm` also caused the refactoring-changes where the whole about-page got removed from `base.jinja` and split into 5 parts stored in `templates/welcome/*`
- As is was convenient from a developer point of view, we added a `WalletsOverviewVm` which has a very similiar set of changes than the `WelcomeVm` in `server_endpoints/wallets`. We also moved the `wallets_api.py` file there. This Change is later used by https://github.com/cryptoadvance/specterext-stacktrack/pull/7
- Parallel to the `WelcomeVm` the `WalletsOverviewVm` also caused the refactorings of `templates/wallet/wallet_overview.jinja` into 4 parts in `templates/wallet/overview`
- An Extension now has a `data_folder` attribute and this folder will get automatically created in `~/.specter/extensions/[extension_id`
* The specter-class has now a splitup between instantiation in its `__init__` method and a new `initialize` method. This splitup also causes the changes in `server.py` which is now instantiating specter via `initialize=False`. Between instantiation and initialising, the `specter.service_manager` attribute gets defined. This change makes it possible that managers which get instantiated within the initialize-method of Specter can call callback-methods via the service_manager. This was eventually not used in https://github.com/cryptoadvance/spectrum/pull/9 but it looked so useful that the change was kept. We'll probably will use it for future extensions.
- a `FlaskThread` is probably not used in Specter directly but maybe useful in extensions
-  
Most of these changes are used by https://github.com/cryptoadvance/spectrum/pull/9

### CODE REVIEW

File: src/cryptoadvance/specter/persistence.py [hunk 1]
```diff
@@ -23,6 +25,67 @@
...
+        else:
+            return f"{self.ext_id}_endpoint"
+
+    @classmethod
+    def from_json(cls, a_dict, *args, **kwargs):
+        """Creates a BusinessObject of the right class"""
+        try:
+            clazz = get_class(a_dict["python_class"])
+        except KeyError:
+            raise SpecterInternalException("dict does not have a python_class")
```
Comment (k9ert) [inline src/cryptoadvance/specter/persistence.py:84]: better error_message

File: src/cryptoadvance/specter/rpc.py [hunk 2]
```diff
@@ -158,6 +158,23 @@ def _detect_rpc_confs_via_env(prefix):
...
+            "password": "",
+            "host": "localhost",
+            "port": "25441",
+            "path": "/svc/spectrum/core_api/",
+            "protocol": "http",
+        }
+    )
+    return rpc_arr
+
+
```
Comment (k9ert) [inline src/cryptoadvance/specter/rpc.py:177]: delete?

File: src/cryptoadvance/specter/managers/service_manager.py [hunk 3]
```diff
@@ -368,8 +379,16 @@ def delete_services_with_unencrypted_storage(self, user: User):
...
+        if self.user_has_encrypted_storage(user=user):
+            # Encrypted Service data is now orphaned since there is no
+            # password. So wipe it from the disk.
+            app.specter.service_encrypted_storage_manager.delete_all_service_data(user)
+            logger.debug(f"Deleted encrypted services for user {user}")
+        else:
+            self.specter.service_unencrypted_storage_manager.delete_all_service_data(
+                user
+            )
+            logger.debug(f"Deleted unencrypted services for user {user}")
```
Comment (k9ert) [inline src/cryptoadvance/specter/managers/service_manager.py:391]: why this changes in this PR?

File: src/cryptoadvance/specter/managers/service_manager.py [hunk 4]
```diff
@@ -368,8 +379,16 @@ def delete_services_with_unencrypted_storage(self, user: User):
         for service_id in services_with_unencrypted_storage:
             self.delete_service_from_user(user, service_id, autosave=True)
 
-        self.specter.service_unencrypted_storage_manager.delete_all_service_data(user)
-        logger.debug(f"Deleted unencrypted services")
+        if self.user_has_encrypted_storage(user=user):
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/managers/service_manager.py:382]: Why is that here?

File: src/cryptoadvance/specter/rpc.py [hunk 5]
```diff
@@ -158,6 +158,23 @@ def _detect_rpc_confs_via_env(prefix):
...
+        {
+            "user": "",
+            "password": "",
+            "host": "localhost",
+            "port": "25441",
+            "path": "/svc/spectrum/core_api/",
+            "protocol": "http",
+        }
+    )
+    return rpc_arr
```
Comment (k9ert) [inline src/cryptoadvance/specter/rpc.py:175]: still needed?

File: src/cryptoadvance/specter/server_endpoints/filters.py [hunk 6]
```diff
@@ -16,6 +17,19 @@ def ascii20(context, name):
     return to_ascii20(name)
 
 
+@pass_context
+@filters_bp.app_template_filter("subrender")
+def subrender_filter(context, value):
+    """based on the idea here:
```
Comment (k9ert) [inline src/cryptoadvance/specter/server_endpoints/filters.py:23]: currently not used but might be usefull, needs documentation as well!

File: src/cryptoadvance/specter/server_endpoints/nodes.py [hunk 7]
```diff
@@ -34,15 +38,25 @@
 def node_settings(node_alias):
     if node_alias:
         try:
-            node = app.specter.node_manager.get_by_alias(node_alias)
+            node: Node = app.specter.node_manager.get_by_alias(node_alias)
+            if not node.is_core_object:
+                return redirect(
```
Comment (k9ert) [inline src/cryptoadvance/specter/server_endpoints/nodes.py:43]: "core" -> naming ?!
consider "specter_core"

File: src/cryptoadvance/specter/server_endpoints/nodes.py [hunk 8]
```diff
@@ -34,15 +38,25 @@
...
+                    url_for(
+                        # This is a convention which should be documented
```
Comment (k9ert) [inline src/cryptoadvance/specter/server_endpoints/nodes.py:45]: This is a convention which should be documented

File: src/cryptoadvance/specter/server_endpoints/welcome/welcome.py [hunk 9]
```diff
@@ -46,6 +48,22 @@ def index():
...
     notify_upgrade(app, flash)
+    # The execute_ext_callbacks method is not really prepared for the things we're doing here.
+    # that's why we need so many lines for just expressing:
+    # "Here is a ViewModel, adjust it if you want"
+    # We need to change that method to enable "middleware"
+    welcome_vm_dict = app.specter.service_manager.execute_ext_callbacks(
+        adjust_view_model, WelcomeVm()
+    )
+    if len(welcome_vm_dict.values()) > 1:
+        raise logger.error("Seems that we have more than one Welcome Extension")
```
Comment (k9ert) [inline src/cryptoadvance/specter/server_endpoints/welcome/welcome.py:59]: raise logger??

File: src/cryptoadvance/specter/server_endpoints/wallets/wallets.py [hunk 10]
```diff
@@ -4,23 +4,25 @@
 from functools import wraps
 
 import requests
+from cryptoadvance.specter.server_endpoints.wallets.wallets_vm import WalletsOverviewVm
```
Comment (k9ert) [inline src/cryptoadvance/specter/server_endpoints/wallets/wallets.py:7]: should be relative import?!

**Comment by moneymanolis** [comment]
### Tasks, quirks, issues
Click to configure results in:
grafik

Shutting down Specter with no node connection in place results in the logs in:
`AttributeError: 'NoneType' object has no attribute 'stop_bitcoind'`
grafik

File: docs/extensions/nodes.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
+# Adding Nodes or NodeTypes
+
+The whole programming-model is based on the Bitcoin-Core API. So we need Bitcoin Core Nodes or Elements Nodes or at least something which behaves like that. So for the Spektrum Integration, we made Extending the Node possible. This is a short description og how that has been done and which extension-points might be helpfull here.
+
+So To Create your own Node, Derive from `AbstractNode`., you have to specify the modules containing subclasses of `Device` in `service.py`:
```
Comment (moneymanolis) [inline docs/extensions/nodes.md:5]: I can't follow here, did you really mean to refer to Device here? In Spectrum's `service.py` for example, there is no link to any Device related module.

File: src/cryptoadvance/specter/util/migrations/migration_0002.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        return """Node-class migration:
+            We introduced the Spectrum Node on an extension. With that, we made the choice of 
+            the Node to be instantiated much more flexible. The node.json get a member called
+            node_class which is the fully qualified package name of the class the NodeManager
+            should instantiate.
+            This migrates all the node.json files to the new format.
+            Effectively it will:
+            * Iterate over all nodes in ~/.specter/nodes/*.json
+            * load each node.json and adds the correct node_class 
+            * stores it again
```
[the remark this answers is not in the document]
Comment (k9ert) [reply src/cryptoadvance/specter/util/migrations/migration_0002.py]: Only removing the `python_class` key is not enough. One need to reestablish the `external_node` as well if needed.

**Review by moneymanolis — APPROVED** [review]

**Comment by moneymanolis** [comment]
- [ ] Issue when trying to update name of the node:

grafik
